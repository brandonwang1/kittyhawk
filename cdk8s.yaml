language: typescript
app: node main.js
imports:
  - k8s 
  - https://github.com/jetstack/cert-manager/releases/download/v1.0.2/cert-manager.yaml
  # The following command was also run:
  # cdk8s import k8s --language typescript --include io.k8s.api.autoscaling.v2beta2.HorizontalPodAutoscaler