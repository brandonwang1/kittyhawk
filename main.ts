import { Construct } from 'constructs';
import { App, Chart } from 'cdk8s';
import { Application, ApplicationOptions } from './lib/application'
import { values } from "./values";
import { CronJob, CronJobOptions } from './lib/cronjob';

export interface Options {

  /**
  * Name of the release.
  * 
  */
  readonly release_name: string;

  /**
  * For compatibility purposes.
  * 
  * @default "0.0.0"
  */
  readonly deploy_version?: string;

  /**
  * What image tag to fetch.
  * 
  * @default "latest"
  */
  readonly image_tag?: string;

  /**
  * List of applications.
  */
  readonly applications: { [x: number]: ApplicationOptions };

  /**
  * List of cronjobs.
  * 
  * @default []
  */
  readonly cronjobs?: { [x: number]: CronJobOptions };

}


export class MyChart extends Chart {
  constructor(scope: Construct, name: string, props: Options) {
    super(scope, name);

    const release_name = props.release_name;
    const applications = props.applications;
    const cronjobs = props.cronjobs || [];

    // Loop through and generate each of the applications
    for (const options in applications) {
      new Application(this, `${release_name}-${applications[options].name}`, applications[options]);
    }

    // Loop through and generate each of the cronjobs
    for (const options in cronjobs) {
      new CronJob(this, `${release_name}-${cronjobs[options].name}`, cronjobs[options]);
    }

  }
}

const app = new App();
new MyChart(app, 'kittyhawk', values);
app.synth();
