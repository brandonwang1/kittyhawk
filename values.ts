import { Options } from './main';

// import { Yaml } from "cdk8s";
// No run time type-checking :(
// export const values1 : Options = Yaml.load("./samples/ohq.yaml")[0] 

export const values : Options =
{
    release_name: "penn-clubs",
    deploy_version: "0.1.22",
    image_tag: "latest", // dynamically fetch this somehow??
    applications: [
        {
            name: 'redis',
            image: 'redis',
            port: 6379
        },
        {
            name: 'django-wgsi',
            image: 'pennlabs/penn-clubs-backend',
            secret: 'penn-clubs',
            replicas: 3,
            extraEnv: [
                { name: 'DOMAIN', value: "pennclubs.com" },
                { name: "DJANGO_SETTINGS_MODULE", value: "pennclubs.settings.production" }
            ],
            ingress: {
                hosts: [{ host: 'pennclubs.com', paths: ["/apis"] }]
            }
        }],
    cronjobs: [{
        name: "rank-clubs",
        schedule: "0 8 * * *",
        image: "pennlabs/penn-clubs-backend",
        secret: "penn-clubs",
        cmd: ["python", "manage.py", "rank"],
    },
    {
        name: "daily-notifications",
        schedule: "0 13 * * *",
        image: "pennlabs/penn-clubs-backend",
        secret: "penn-clubs",
        cmd: ["python", "manage.py", "daily_notifications"],
    }]
}

// console.log(JSON.stringify(values1))

