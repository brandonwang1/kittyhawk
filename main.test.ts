import {MyChart} from './main';
import {Testing} from 'cdk8s';
import { values } from './values';

describe('Placeholder', () => {
  test('Empty', () => {
    const app = Testing.app();
    const chart = new MyChart(app, 'test-chart', values);
    const results = Testing.synth(chart)
    expect(results).toMatchSnapshot();
  });
});
