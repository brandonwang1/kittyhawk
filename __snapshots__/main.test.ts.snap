// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder Empty 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-chart-penn-clubs-redis-service-redis-d9bf58c2",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
          "targetPort": 6379,
        },
      ],
      "selector": Object {
        "name": "penn-clubs-redis",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-chart-penn-clubs-redis-deployment-redis-24f84422",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "penn-clubs-redis",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "penn-clubs-redis",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "redis",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-chart-penn-clubs-django-wgsi-service-django-wgsi-93922c6f",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "penn-clubs-django-wgsi",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-cha-penn-clubs-django-wgsi-deployment-django-wgsi-e753a545",
    },
    "spec": Object {
      "replicas": 3,
      "selector": Object {
        "matchLabels": Object {
          "name": "penn-clubs-django-wgsi",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "penn-clubs-django-wgsi",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "pennlabs/penn-clubs-backend",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "name": "test-chart-penn-clubs-django-wgsi-ingress-django-wgsi-e17422bd",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "pennclubs.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "penn-clubs-django-wgsi",
                  "servicePort": 80,
                },
                "path": "/apis",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "pennclubs.com",
          ],
          "secretName": "pennclubs.com-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "name": "test-ch-penn-clubs-django-wgsi-certificate-django-wgsi-cfcf92a3",
    },
    "spec": Object {
      "dnsNames": Array [
        "",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "abc-tls",
    },
  },
  Object {
    "apiVersion": "batch/v1beta1",
    "kind": "CronJob",
    "metadata": Object {
      "name": "test-chart-penn-clubs-rank-clubs-cronjob-rank-clubs-6e6eb51f",
    },
    "spec": Object {
      "failedJobsHistoryLimit": 1,
      "jobTemplate": Object {},
      "schedule": "0 8 * * *",
      "successfulJobsHistoryLimit": 1,
    },
  },
  Object {
    "apiVersion": "batch/v1beta1",
    "kind": "CronJob",
    "metadata": Object {
      "name": "penn-clubs-daily-notificat-cronjob-daily-notifications-dafff798",
    },
    "spec": Object {
      "failedJobsHistoryLimit": 1,
      "jobTemplate": Object {},
      "schedule": "0 13 * * *",
      "successfulJobsHistoryLimit": 1,
    },
  },
]
`;
